pkg/agent/handler/credential.go

```diff
@@ -194,67 +234,164 @@ func (h *credentials) getCRD(ctx context.Context, cred *mv1.Credential) (*mv1.Cr
-			if err != nil {
+	// provision external client
+	resClientMetadata, err := p.RegisterClient(clientMetadata)
+	if err != nil {
+		return err
+	}
 
-				log.Error(err)
-				continue
-			}
+	cr.idpCredData.clientID = resClientMetadata.GetClientID()
+	cr.idpCredData.clientSecret = resClientMetadata.GetClientSecret()
+	return nil
+}
 
-			data[key] = base64.StdEncoding.EncodeToString([]byte(str))
-		}
+func (h *credentials) unregisterIDPClientCredential(cr *provCreds) error {
+	p := cr.GetIDPProvider()
+	err := p.UnregisterClient(cr.idpCredData.GetClientID())
+	if err != nil {
+		return err
 	}
 
-	return data
+	cr.idpCredData.clientID = cr.idpCredData.GetClientID()
+	return nil
+}
+
+func (h *credentials) getProvisionedCredentialData(provCreds *provCreds, credentialData prov.Credential) prov.Credential {
+	if provCreds.IsIDPCredential() {
+		return prov.NewCredentialBuilder().SetOAuthIDAndSecret(
+			provCreds.GetIDPCredentialData().GetClientID(),
+			provCreds.GetIDPCredentialData().GetClientSecret(),
+		)
+	}
+	return credentialData
 }
 
 type provCreds struct {
 	managedApp  string
 	credType    string
 	id          string
+	name        string
 	credData    map[string]interface{}
 	credDetails map[string]interface{}
 	appDetails  map[string]interface{}
+	idpCredData *idpCredData
+	idpProvider oauth.Provider
+}
+
+type idpCredData struct {
+	clientID        string
+	clientSecret    string
+	scopes          []string
+	grantTypes      []string
+	tokenAuthMethod string
+	responseTypes   []string
+	redirectURLs    []string
+	jwksURI         string
+	publicKey       string
 }
 
-func newProvCreds(cr *mv1.Credential, appDetails map[string]interface{}) *provCreds {
+func newProvCreds(cr *mv1.Credential, appDetails map[string]interface{}, provData map[string]interface{}, idpProviderRegistry oauth.ProviderRegistry) (*provCreds, error) {
 	credDetails := util.GetAgentDetails(cr)
 
-	return &provCreds{
+	provCred := &provCreds{
 		appDetails:  appDetails,
 		credDetails: credDetails,
 		credType:    cr.Spec.CredentialRequestDefinition,
 		credData:    cr.Spec.Data,
 		managedApp:  cr.Spec.ManagedApplication,
 		id:          cr.Metadata.ID,
+		name:        cr.Name,
+	}
+
+	// Setup external credential request data to be used for provisioning
+	if idpTokenURL, ok := provCred.credData[prov.IDPTokenURL].(string); ok {
+		p, err := idpProviderRegistry.GetProviderByTokenEndpoint(idpTokenURL)
+		if err != nil {
+			return nil, fmt.Errorf("IDP provider not found for credential request")
+		}
+		provCred.idpProvider = p
+		provCred.idpCredData = newIDPCredData(p, provCred.credData, provData)
 	}
+
+	return provCred, nil
+}
+
+// newIDPCredData - reads the idp client metadata from credential request
+func newIDPCredData(p oauth.Provider, credData, provData map[string]interface{}) *idpCredData {
+	cd := &idpCredData{}
+
+	if provData != nil {
+		if data, ok := provData[prov.OauthClientID]; ok && data != nil {
+			cd.clientID = data.(string)
```

Comment:
Maybe check if the type conversion was ok

Reply:
It's up to you. There are a few places below here too that don't check for ok. But maybe it is fine.

Reply:
Sure, will do that change